sess_fail_* and vcp stats, holddown for vcp

nigoroll opened the pull request:
This should address all the issues documented in #2622 
* visibility for accept and vcp errors
* holddown for vcp errors

Dridi commented on bin/varnishd/cache/cache_tcp_pool.c:
```diff
@@ -240,8 +247,26 @@ VCP_New(struct conn_pool *cp, const void *id, void *priv,
+		AN(addr);
+		VTCP_name(addr, abuf, sizeof abuf, pbuf, sizeof pbuf);
+		cp->stats = VSC_vcp_New(NULL, &cp->vsc, "%p.%s:%s", id,
+					abuf, pbuf);
```
Until Varnish 4.0 backend counters used to include the inet4/inet6/port tuple in the name, for example:

```
VBE.default(127.0.0.1,::1,8080).*
```

I think we should inherit that behavior for the new counters.

Dridi commented on bin/varnishd/VSC_vcp.vsc:
```diff
@@ -0,0 +1,120 @@
+.. varnish_vsc:: n_conn
+	:type:	counter
+	:level:	diag
+	:oneliner:	Number of cached connections
```
s/cache/pooled/ ?

Dridi commented on bin/varnishd/cache/cache_tcp_pool.c:
```diff
@@ -347,27 +350,52 @@ VCP_Open(struct conn_pool *cp, double tmo, const void **privp)
 
+	while (cp->holddown > 0) {
+		Lck_Lock(&cp->mtx);
+		if (cp->holddown == 0) {
```
How can it be `== 0` if you need `> 0` to enter the loop?

Should it be this instead? `while (cp->holddown >= 0) {`

nigoroll commented:
@Dridi  thank you for your feedback, see last two added commits

Dridi commented on bin/varnishd/cache/cache_tcp_pool.c:
```diff
@@ -279,32 +307,121 @@ VCP_Rel(struct conn_pool *cp)
+
+	while (cp->holddown > 0) {
+		Lck_Lock(&cp->mtx);
+		if (cp->holddown == 0) {
```
This `if` statement is still inconsistent with the `while` above.

nigoroll replied:
github had shown me your initial comment on this as outdated, so I thought you had taken it back.

So what is happening here is that holddown can be modified concurrently to a thread in open. We want to avoid taking the lock each time, so we accept to miss a holddown in favor of somehow less locking overhead. But for the reverse case that the holddown has been cleared concurrently, we want to ensure that we actually see it cleared.

We we check unprotected first but then, under the lock, we make sure that holddown has not been cleared concurrently.

nigoroll replied:
BTW, the main motivation for the lock is to get a consistent view on holddown and holddown_errno. While we could juggle membars, a simple mutex is easier and safe and with the above optimization the whole thing has close to no overhead for the base case (no connection errors).

Dridi replied:
I cleaned my glasses and finally was able to see that glaring lock in-between. Thank you for your patience :)

nigoroll commented:
rebased & force pushed

nigoroll commented:
the first patch `sess_fail_*` was accepted by phk and is now merged as 3f4cd9297d53b1e18e594c83d9fd068d4dd4a666

I've squashed and rebased the other patches, leaving us with three commits.

To summarize today's irc discussion:
* I have implemented the backend connection errors on the VCP layer for these reasons:
  * this is where the connection `open()` happens
  * we already had other statistics about pools which we did not expose yet could be useful
  * connections are shared across backends and vcls
* yet an argument can be made to expose the open error counters on the backend (VBE) layer, per backend (with semantics _number of errors this backend experienced when it happend not to be able to reuse a recycled connection_), as this avoids involving users in another layer of indirection.
  * if we went this route, some counters would remain on the pool side which I think should still be exposed
* Another option we discussed are _VSC symlinks_ to link backends to pools (won't work when we start using different VCP keys)

@bsdphk please make a decision and I will be happy to change the patches accordingly

bsdphk commented:
Decision:  Counters go with other backen counters.

The postulated clarity of having them on the pool is too speculative to warrant the complexity required at this point in time.

nigoroll commented:
I will update the PR

nigoroll commented:
Continued in #2699
